### PR TITLE
Frontend Create Backfill

### PR DESCRIPTION
**Pull request by aLekSer**
Initial version of CreateBackfill. It stores an empty regular TicketIds list.

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
Wait on StateStare changes.

### CODE REVIEW

File: internal/app/frontend/frontend_service.go
Comment (hsorellana) [inline internal/app/frontend/frontend_service.go:134]: There's a missing closing curly brace here

Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:134]: That's true, I cannot run the test without state store changes.

File: internal/api/internal.proto
Comment (scosgrave) [inline internal/api/internal.proto:1]: I think `statestore.proto` might be a more appropriate name for this file. It matches the style of using the module name in the proto file name that `synchronizer.proto` has started.

Comment (Laremere) [reply internal/api/internal.proto:1]: messages.proto would also be appropriate, as we put all message definitions that aren't request/response types in a similarly named file in the public api.

Comment (Laremere) [reply internal/api/internal.proto:1]: Oh that file is actually being added by https://github.com/googleforgames/open-match/pull/1273

Comment (scosgrave) [reply internal/api/internal.proto:1]: My initial take on how this file should be named was not right. Laramere educated me on how these files are named. The naming pattern is not based on the module name. Instead, services and their req/resp usually go into their own proto file, and objects that aren't coupled to a service's API usually go in a messages.proto file. 
+1 for messages.proto

Comment (aLekSer) [reply internal/api/internal.proto:1]: I will copy a following thread back to the author @akremsa

File: internal/app/frontend/frontend_service.go
Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:121]: This is an error to have not nil createTime:
```
		return nil, status.Errorf(codes.InvalidArgument, "backfills cannot be created with create time set")
```

File: internal/app/frontend/frontend_service.go
Comment (akremsa) [inline internal/app/frontend/frontend_service.go:101]: why don't you check req for nil?

File: internal/app/frontend/frontend_service_test.go
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:381]: require.NotNil(t, backfill)

File: internal/testing/e2e/backfill_test.go
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:43]: why do you mix assert and require? 
please use require

File: internal/testing/e2e/backfill_test.go
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:49]: please use common naming like **result** or **actual**, it will be easier to read this code.

File: internal/testing/e2e/backfill_test.go
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:36]: require.NotNil(t, b1)

File: internal/testing/e2e/backfill_test.go
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:39]: require.NotNil(t, b2)

File: internal/testing/e2e/backfill_test.go
Comment (akremsa) [inline internal/testing/e2e/backfill_test.go:46]: What does it mean? Could you rephrase it, please?

File: internal/app/frontend/frontend_service.go
Comment (akremsa) [inline internal/app/frontend/frontend_service.go:103]: Please test these req validations in e2e tests

File: internal/app/frontend/frontend_service_test.go
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:135]: require.NotNil(t, res)

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: Why do you test doCreateBackfill instead of CreateBackfill?

**Review by akremsa — CHANGES_REQUESTED** [review]
Some minor changes requested

File: internal/app/frontend/frontend_service.go
Comment (hsorellana) [inline internal/app/frontend/frontend_service.go:133]: Before returning, after my PR is merged, here should be a store.IndexBackfill call also

Comment (aLekSer) [reply internal/app/frontend/frontend_service.go:133]: That's expected to have. I will add in a different PR.

File: internal/app/frontend/frontend_service_test.go
Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:135]: I am testing real handlers as part of E2E test.

File: internal/testing/e2e/backfill_test.go
Comment (aLekSer) [reply internal/testing/e2e/backfill_test.go:46]: I have updated this comment, but in a separate branch, sorry. I meant that all fields other then CreateTime and Ids are the same if we CreateBackfill with the same input args.

File: internal/app/frontend/frontend_service_test.go
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:91]: While I think this follows the patterns of the testing for the other calls, I'm unsure about the value these tests are adding:

- I don't think calling with canceled context actually finds problems, at most if checks that one use of context is done correctly, but not that all uses are correct.
- It took a bit to get there, but I'm pretty confident that "preAction" is an anti-pattern at this point.  It's adding too much logic to tests.
- See also, the if err != nil check in the test code is definitely logic.  This should be removed, and is evidence of the tests being over-generalized.

My recommendation:
Get the test coverage using the e2e tests for:
- creating
- getting backfill created
- getting backfill which isn't present, and getting not found
(first two are already done, as far as I can tell)

At most, keep only the canceled context tests here, but I'm fine not having those.
The other tests that follow these patterns for existing features should probably be cleaned up, but that's out of scope for this project.

Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:91]: Thanks for your feedback, will address all comments

File: internal/app/frontend/frontend_service_test.go
Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: This is a unit test, it has nothing to do with e2e one. _doCreateBackfill_ is not an entry point to the current package. Please, test exported methods only, as you do in TestDoGetBackfill.

Comment (aLekSer) [reply internal/app/frontend/frontend_service_test.go:135]: First of all we follow the same pattern for `TestDoCreateTicket()`, second Scott asked me to use E2E to cover those handlers:
https://github.com/googleforgames/open-match/pull/1279/files#r528961563

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go:135]: If there is the same approach in TestDoCreateTicket it doesn't mean that it is done right. You should not rely on any other tests (e2e, etc) while working on unit tests. Here you **ignore** everything which comes before _doCreateBackfill_.

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go]: @Laremere do you mean that we should test public API in e2e tests, rather than unit ones?
Because at this moment I don't see a clear boundary between tickets e2e and tickets unit tests, for instance.
If so - there is no point to add and support tests in _frontend_ package right now.

Comment (akremsa) [reply internal/app/frontend/frontend_service_test.go]: With _TestDoCreateTickets_ half of a method is tested in frontend package, another part (validations) in e2e, which is a bit confusing.

**Review by akremsa — APPROVED** [review]

File: internal/app/frontend/frontend_service_test.go
Comment (akremsa) [inline internal/app/frontend/frontend_service_test.go:152]: Ok, now it's better. But why do you retain TestDoCreateTickets now?

**Comment by aLekSer** [comment]
Hello @Laremere I have applied all comments, if I did something wrong please leave feedback.

File: internal/app/frontend/frontend_service_test.go
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:153]: Avoid clauses like this in tests: instead this should be two separate test cases.

File: internal/app/frontend/frontend_service_test.go
Comment (Laremere) [inline internal/app/frontend/frontend_service_test.go:164]: Should be in a separate test case.

File: internal/testing/e2e/backfill_test.go
Comment (Laremere) [inline internal/testing/e2e/backfill_test.go:64]: I think this should be a separate test case?

**Review by Laremere — APPROVED** [review]
Adding a few comments, don't want to block this on nit for testing, so merging with comments.  Cleanup to match my desired state doesn't involve only the new tests here, so it's fine.

File: internal/testing/e2e/backfill_test.go
Comment (aLekSer) [reply internal/testing/e2e/backfill_test.go:64]: OK, I will make it as a separate PR.

**Pull request merged**
